Client/Frontend/Browser/Onboarding/OnboardingAdsCountdownView.swift

```diff
@@ -0,0 +1,370 @@
+            
+            countdownLayer.foregroundColor = UX.gradientPurpleLight.cgColor
+            shapeLayer.strokeColor = UX.backgroundGrayLight.cgColor
+            strokeBallLayer.backgroundColor = UX.gradientOrangeLight.cgColor
+        }
+    }
+    
+    @available(*, unavailable)
+    required init(coder: NSCoder) { fatalError() }
+    
+    func animate(from startOffset: CGFloat, to endOffset: CGFloat, duration: CFTimeInterval, completion: (() -> Void)? = nil) {
+        CATransaction.begin()
+        
+        countdownLayer.string = String(Int(duration)) as NSString
+        
+        let backgroundAnimation = CABasicAnimation(keyPath: "strokeStart")
+        backgroundAnimation.duration = duration
+        backgroundAnimation.isRemovedOnCompletion = false
+        backgroundAnimation.fromValue = startOffset
+        backgroundAnimation.toValue = endOffset
+        backgroundAnimation.fillMode = .forwards
+        backgroundAnimation.timingFunction = CAMediaTimingFunction(name: .linear)
+        
+        let strokeAnimation = CABasicAnimation(keyPath: "strokeEnd")
+        strokeAnimation.duration = duration
+        strokeAnimation.isRemovedOnCompletion = false
+        strokeAnimation.fromValue = startOffset
+        strokeAnimation.toValue = endOffset
+        strokeAnimation.fillMode = .forwards
+        strokeAnimation.timingFunction = CAMediaTimingFunction(name: .linear)
+        
+        let ballAnimation = CAKeyframeAnimation(keyPath: "position")
+        ballAnimation.duration = duration
+        ballAnimation.isRemovedOnCompletion = false
+        ballAnimation.calculationMode = .paced
+        ballAnimation.path = createAnimationPath().cgPath
+        ballAnimation.fillMode = .forwards
+        ballAnimation.timingFunction = CAMediaTimingFunction(name: .linear)
+        
+        let countdownAnimation = createCountdownAnimation(duration: duration, startValue: Int(duration), endValue: 0)
+        
+        CATransaction.setCompletionBlock({
+            self.strokeLayer.strokeEnd = endOffset
+            self.shapeLayer.strokeStart = endOffset
+            self.countdownLayer.string = String(Int(0))
+            
+            self.shapeLayer.removeAnimation(forKey: "backgroundAnimation")
+            self.strokeLayer.removeAnimation(forKey: "strokeAnimation")
+            self.strokeBallLayer.removeAnimation(forKey: "ballAnimation")
+            self.countdownLayer.removeAnimation(forKey: "countdownAnimation")
+            
+            completion?()
+        })
+        
+        shapeLayer.add(backgroundAnimation, forKey: "backgroundAnimation")
+        strokeLayer.add(strokeAnimation, forKey: "strokeAnimation")
+        strokeBallLayer.add(ballAnimation, forKey: "ballAnimation")
+        countdownLayer.add(countdownAnimation, forKey: "countdownAnimation")
+        
+        CATransaction.commit()
+    }
+}
+
+private class CenteredTextLayer: CATextLayer {
```

Comment:
Can't animate UILabel's text property so I can't animate the countdown. If I use `CATextLayer`, I can animate the string property

Reply:
What are we animating? I don't see any animation of the actual text in simulator, it just goes from 3 to 2 instantly? And its blurry like @danishjafri88 mentioned probably because it's a CATextLayer.  Unless I'm missing something, lets just replace it with a UILabel

Reply:
Blurriness is fixed by setting the scale on the layer.

To keep things in sync with the rest of animations, it must be a CATextLayer. A UILabel's text cannot be animated so it won't be able to countdown with the rest of the animation, unless I add a timer and then manually change the text every tick.

Plus when timers go in the background they get expired so it's even worse than the current solution which works fine.